From 61b8596727cb75c134ca6367b956d0e4596bfc69 Mon Sep 17 00:00:00 2001
From: Tallapragada Kalyan <quic_ktallapr@quicinc.com>
Date: Thu, 9 Jun 2022 09:32:38 +0530
Subject: [PATCH 179/500] net: set skb's fast_xmit flag in dev_fast_xmit API

set skb's fast_xmit flag in dev_fast_xmit API for linear packets
WiFi tx path can avoid some overhead due to checks based on this flag

Change-Id: Ied29f9d615d0cf48dd9dcd7fcf0fb210eb259a8f
Signed-off-by: Tallapragada Kalyan <quic_ktallapr@quicinc.com>
---
 include/linux/skbuff.h | 2 ++
 net/core/dev.c         | 4 ++++
 2 files changed, 6 insertions(+)

diff --git a/include/linux/skbuff.h b/include/linux/skbuff.h
index 5f681e154b23..48205abb29f5 100644
--- a/include/linux/skbuff.h
+++ b/include/linux/skbuff.h
@@ -989,6 +989,8 @@ struct sk_buff {
 #if IS_ENABLED(CONFIG_IP_SCTP)
 	__u8			csum_not_inet:1;
 #endif
+	/* Linear packets processed by dev_fast_xmit() */
+	__u8			fast_xmit:1;
 
 #if defined(CONFIG_NET_SCHED) || defined(CONFIG_NET_XGRESS)
 	__u16			tc_index;	/* traffic control index */
diff --git a/net/core/dev.c b/net/core/dev.c
index 8c922da91cf9..d65cb38b9e99 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -4269,6 +4269,10 @@ bool dev_fast_xmit(struct sk_buff *skb,
 	int cpu;
 	netdev_tx_t rc;
 
+	/* the fast_xmit flag will avoid multiple checks in wifi xmit path */
+	if (likely(!skb_is_nonlinear(skb)))
+		skb->fast_xmit = 1;
+
 	if (unlikely(!(dev->flags & IFF_UP))) {
 		return false;
 	}
-- 
2.34.1

